### ISSUE TITLE
hostname does not match the server certificate

### ISSUE THREAD

**Issue by servomac** [newrelic/centurion#113]
Hi!

I have configured a server to use HTTPS on the docker API. I have added the certificates on ~/.docker/, and I can connect using docker. As an example:

> tpiza@spring:~/projects/mobilestatus$ docker --tlsverify -H=sheriff.domain.com:2376 ps
>CONTAINER ID        IMAGE                                      COMMAND                CREATED             STATUS              PORTS                           NAMES
>3e0adaab3d21        domain/humhub:latest                  "/apache-start.sh"     2 days ago          Up 15 minutes       443/tcp, 0.0.0.0:8081->80/tcp   humhub-humhub       
>497965b559cb        domain/mariadb-replicator-db:latest   "/docker-entrypoint.   2 days ago          Up 3 hours          3306/tcp                        humhub-mariadb      
>1d45231852e0        tianon/exim4:latest                        "entrypoint.sh exim    2 days ago          Up 3 hours          25/tcp                          exim4      

But i'm completely unable to deploy/list anything using centurion 1.6.0. My configuration:

>tpiza@spring:~/projects/mobilestatus$ cat config/centurion/redis.rake 
>namespace :environment do
>  task :common do
>    set :image, 'library/redis'
>    set :name, 'redis'
>  end
>
>  desc 'Staging environment'
>  task :staging => :common do
>    set_current_environment(:staging)
>    host '192.168.1.205:2375'
>    host_port 16379, container_port: 6379
>  end
>
>  desc 'Production environment'
>  task :production => :common do
>    set_current_environment(:production)
>    set :tlsverify, true
>    host 'sheriff.domain.com:2376'
>    host_port 6379, container_port: 6379
>  end
>end

And the error (obviously I have generated the certs specifying the correct hostname, otherwise it will not work when i connect directly using the docker client :S):

>tpiza@spring:~/projects/mobilestatus$ centurion -p redis -e production -a list
>** Invoke environment:production (first_time)
>** Invoke environment:common (first_time)
>** Execute environment:common
>** Execute environment:production
>** Invoke centurion:setup (first_time)
>** Invoke centurion:clean_environment (first_time)
>** Execute centurion:clean_environment
>** Execute centurion:setup
>** Invoke list (first_time)
>** Execute list
>** Invoke list:tags (first_time)
>** Execute list:tags
>GET: "https://library/v1/repositories/redis/tags"
>E, [2015-05-14T17:43:12.975252 #1199] ERROR -- : Couldn't communicate with Registry: getaddrinfo: >Name or service not known (SocketError)
>
>** Invoke list:running_containers (first_time)
>** Execute list:running_containers
>I, [2015-05-14T17:43:12.975549 #1199]  INFO -- : ----- Connecting to Docker on sheriff.domain.com -----
>E, [2015-05-14T17:43:13.168049 #1199] ERROR -- : Couldn't communicate with Docker on sheriff.domain.com: hostname does not match the server certificate (OpenSSL::SSL::SSLError)
>/usr/lib/ruby/1.9.1/openssl/ssl-internal.rb:129:in `post_connection_check': hostname does not match the server certificate (OpenSSL::SSL::SSLError) (Excon::Errors::SocketError)
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/ssl_socket.rb:137:in `initialize'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/connection.rb:387:in `new'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/connection.rb:387:in `socket'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/connection.rb:106:in `request_call'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/middlewares/mock.rb:47:in `request_call'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/middlewares/instrumentor.rb:22:in `request_call'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/middlewares/base.rb:15:in `request_call'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/middlewares/base.rb:15:in `request_call'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/middlewares/base.rb:15:in `request_call'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/connection.rb:233:in `request'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon.rb:229:in `get'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/docker_via_api.rb:19:in `ps'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/tasks/list.rake:45:in `block (3 levels) in <top (required)>'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/deploy_dsl.rb:6:in `call'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/deploy_dsl.rb:6:in `block (2 levels) in on_each_docker_host'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/docker_server_group.rb:22:in `call'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/docker_server_group.rb:22:in `block in each'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/docker_server_group.rb:20:in `each'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/docker_server_group.rb:20:in `each'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/deploy_dsl.rb:6:in `block in on_each_docker_host'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/deploy_dsl.rb:6:in `tap'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/centurion/deploy_dsl.rb:6:in `on_each_docker_host'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/tasks/list.rake:43:in `block (2 levels) in <top (required)>'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `call'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `block in execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `each'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:158:in `block in invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/monitor.rb:211:in `mon_synchronize'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:151:in `invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:144:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/capistrano_dsl.rb:88:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/tasks/list.rake:5:in `block in <top (required)>'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `call'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `block in execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `each'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:158:in `block in invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/monitor.rb:211:in `mon_synchronize'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:151:in `invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:144:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/lib/capistrano_dsl.rb:88:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.6.0/bin/centurion:80:in `<top (required)>'
>	from /usr/bin/centurion:23:in `load'
>	from /usr/bin/centurion:23:in `<main>'

Did I have any mistake in my config/environment, or there is any real problem? Thanks a lot for your time!

**Comment by benders** [comment]
Fascinating! I can totally see how this would be causing trouble though. It looks like perhaps the pure HTTP parts aren't picking up the CA Cert needed. Did that same config work with an earlier version of Centurion, or are you starting fresh?

**Comment by servomac** [comment]
It's my first time :-)

**Comment by servomac** [comment]
I have been trying centurion 1.5.0; I have modified the config/centurion/redis.rake to adapt the tls directives to 1.5.0 (:tlsverify now is :tls), but is not working yet.. I have also modified the docker server to listen on the default port. Docker client keeps ok, but centurion is still unable to connect (and prints an error specifying that *Couldn't communicate with Docker on sheriff.domain.com: end of file reached (EOFError)*):

> tpiza@spring:~/projects/mobilestatus$ docker --tlsverify -H=tcp://sheriff.domain.com:2375 version
>Client version: 1.6.0
>Client API version: 1.18
>Go version (client): go1.4.2
>Git commit (client): 4749651
>OS/Arch (client): linux/amd64
>Server version: 1.6.1
>Server API version: 1.18
>Go version (server): go1.4.2
>Git commit (server): 97cd073
>OS/Arch (server): linux/amd64

Error with centurion 1.5.0:

>tpiza@spring:~/projects/mobilestatus$ centurion -p redis -e production -a list
>** Invoke environment:production (first_time)
>** Invoke environment:common (first_time)
>** Execute environment:common
>** Execute environment:production
>** Invoke list (first_time)
>** Execute list
>** Invoke list:tags (first_time)
>** Execute list:tags
>GET: "https://library/v1/repositories/redis/tags"
>E, [2015-05-15T10:56:46.487176 #8189] ERROR -- : Couldn't communicate with Registry: getaddrinfo: Name or service not known (SocketError)
>
>** Invoke list:running_containers (first_time)
>** Execute list:running_containers
>I, [2015-05-15T10:56:46.487544 #8189]  INFO -- : ----- Connecting to Docker on sheriff.domain.com -----
>E, [2015-05-15T10:56:46.551442 #8189] ERROR -- : Couldn't communicate with Docker on sheriff.domain.com: end of file reached (EOFError)
>/usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/socket.rb:45:in `read_nonblock': end of file reached (EOFError) (Excon::Errors::SocketError)
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/socket.rb:45:in `readline'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/response.rb:163:in `parse_headers'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/response.rb:75:in `parse'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/middlewares/response_parser.rb:6:in `response_call'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/connection.rb:372:in `response'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon/connection.rb:236:in `request'
>	from /usr/lib/ruby/gems/1.9.1/gems/excon-0.45.3/lib/excon.rb:229:in `get'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/docker_via_api.rb:19:in `ps'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/tasks/list.rake:45:in `block (3 levels) in <top (required)>'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/deploy_dsl.rb:6:in `call'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/deploy_dsl.rb:6:in `block (2 levels) in on_each_docker_host'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/docker_server_group.rb:22:in `call'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/docker_server_group.rb:22:in `block in each'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/docker_server_group.rb:20:in `each'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/docker_server_group.rb:20:in `each'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/deploy_dsl.rb:6:in `block in on_each_docker_host'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/deploy_dsl.rb:6:in `tap'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/centurion/deploy_dsl.rb:6:in `on_each_docker_host'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/tasks/list.rake:43:in `block (2 levels) in <top (required)>'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `call'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `block in execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `each'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:158:in `block in invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/monitor.rb:211:in `mon_synchronize'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:151:in `invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:144:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/capistrano_dsl.rb:88:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/tasks/list.rake:5:in `block in <top (required)>'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `call'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:205:in `block in execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `each'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:200:in `execute'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:158:in `block in invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/monitor.rb:211:in `mon_synchronize'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:151:in `invoke_with_call_chain'
>	from /usr/lib/ruby/1.9.1/rake/task.rb:144:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/lib/capistrano_dsl.rb:88:in `invoke'
>	from /usr/lib/ruby/gems/1.9.1/gems/centurion-1.5.0/bin/centurion:78:in `<top (required)>'
>	from /usr/bin/centurion:23:in `load'
>	from /usr/bin/centurion:23:in `<main>'

Thanks a lot for your time. I can wait to use centurion to deploy in my production environment, but I need to assure a minimum of security :D

**Comment by hugochinchilla** [comment]
Hi, I'm a coworker of @servomac, we have been able to trace the problem down to the excon gem, 
the funny part is that using the IP instead of the hostname does work.

```ruby
require 'rubygems'
require 'excon'

#address = 'https://sheriff.domain.com:2376/info'
address = 'https://xx.xx.xx.xx:2376/info' # using IP works!

Excon.defaults[:connect_timeout] = 120 
Excon.defaults[:read_timeout]    = 120 
Excon.defaults[:write_timeout]   = 120 
Excon.defaults[:debug_request]   = true
Excon.defaults[:debug_response]  = true
Excon.defaults[:nonblock]        = false
Excon.defaults[:tcp_nodelay]     = true
Excon.defaults[:ssl_ca_file]     = File.expand_path('~/.docker/ca.pem')
#Excon.defaults[:ssl_verify_peer] = false

tls_args = {
  :client_cert => File.expand_path('~/.docker/cert.pem'),
  :client_key => File.expand_path('~/.docker/key.pem')
}

response = Excon.get(address, tls_args.merge(:headers => {'Accept' => 'application/json'}))
puts response.body
```

**Comment by hugochinchilla** [comment]
Digging more through the rabbit hole I spotted the problem looking into the OpenSSL library, the method `verify_certificate_identity` inside `openssl/ssl.rb` explains why the exception was happening, it is a problem in the way we emitted the certificate, sadly we just copy&pasted from the Docker documentation at https://docs.docker.com/articles/https. 

The solution is not to set an Subject Alternative Name, or if so, creating one with the DNS name also on it, because if present openssl will not check the Common Name field. But as this problem only occurs from ruby and not with the docker client I think it might be a problem with the openssl implementation in ruby.
